desc/compute/_equil.py

```diff
@@ -4,118 +4,423 @@
-    """Compute force error components."""
-    data = compute_pressure(
-        params,
-        transforms,
-        profiles,
-        data=data,
-        **kwargs,
-    )
-    data = compute_contravariant_current_density(
-        params,
-        transforms,
-        profiles,
-        data=data,
-        **kwargs,
-    )
-    data = compute_contravariant_metric_coefficients(
-        params,
-        transforms,
-        profiles,
-        data=data,
-        **kwargs,
-    )
-    data = compute_geometry(
-        params,
-        transforms,
-        profiles,
-        data=data,
-        **kwargs,
+from .data_index import register_compute_fun
+from .utils import dot
+
+
+@register_compute_fun(
+    name="J^rho",
+    label="J^{\\rho}",
+    units="A \\cdot m^{-3}",
+    units_long="Amperes / cubic meter",
+    description="Contravariant radial component of plasma current density",
+    dim=1,
+    params=[],
+    transforms={},
+    profiles=[],
+    function_of="rtz",
+    data=["sqrt(g)", "B_zeta_t", "B_theta_z"],
+)
+def _J_sup_rho(params, transforms, profiles, data, **kwargs):
+    data["J^rho"] = (data["B_zeta_t"] - data["B_theta_z"]) / (mu_0 * data["sqrt(g)"])
+    return data
+
+
+@register_compute_fun(
+    name="J^theta",
+    label="J^{\\theta}",
+    units="A \\cdot m^{-3}",
+    units_long="Amperes / cubic meter",
+    description="Contravariant poloidal component of plasma current density",
+    dim=1,
+    params=[],
+    transforms={},
+    profiles=[],
+    function_of="rtz",
+    data=["sqrt(g)", "B_rho_z", "B_zeta_r"],
+)
+def _J_sup_theta(params, transforms, profiles, data, **kwargs):
+    data["J^theta"] = (data["B_rho_z"] - data["B_zeta_r"]) / (mu_0 * data["sqrt(g)"])
+    return data
+
+
+@register_compute_fun(
+    name="J^zeta",
+    label="J^{\\zeta}",
+    units="A \\cdot m^{-3}",
+    units_long="Amperes / cubic meter",
+    description="Contravariant toroidal component of plasma current density",
+    dim=1,
+    params=[],
+    transforms={},
+    profiles=[],
+    function_of="rtz",
+    data=["sqrt(g)", "B_theta_r", "B_rho_t"],
+)
+def _J_sup_zeta(params, transforms, profiles, data, **kwargs):
+    data["J^zeta"] = (data["B_theta_r"] - data["B_rho_t"]) / (mu_0 * data["sqrt(g)"])
+    return data
+
+
+@register_compute_fun(
+    name="J",
+    label="\\mathbf{J}",
+    units="A \\cdot m^{-3}",
+    units_long="Amperes / cubic meter",
+    description="Plasma current density",
+    dim=3,
+    params=[],
+    transforms={},
+    profiles=[],
+    function_of="rtz",
+    data=["J^rho", "J^theta", "J^zeta", "e_rho", "e_theta", "e_zeta"],
+)
+def _J(params, transforms, profiles, data, **kwargs):
+    data["J"] = (
+        data["J^rho"] * data["e_rho"].T
+        + data["J^theta"] * data["e_theta"].T
+        + data["J^zeta"] * data["e_zeta"].T
+    ).T
+    return data
+
+
+@register_compute_fun(
+    name="J_R",
+    label="J_{R}",
+    units="A \\cdot m^{-3}",
```

Comment:
I am pretty sure this should be A / m^2, the contravariant components J^rho etc yes are A/m^3, but when you go to lab frame you multiple J^rho by e_rho etc which then gives A/m^3. Besides, physically current density is A/m^2